### ISSUE TITLE
can emitFirrtl but not emitVerilog

### ISSUE THREAD

**Issue by makslevental** [chipsalliance/firrtl#2282]
### What is the current behavior?

can run

```
object LineBuffer extends App {
  val dWidth = 16
  (new ChiselStage).emitFirrtl(
    new LineBuffer(dtype = UInt(dWidth.W), colsIn = 3, rowsOut = 3),
    args
  )
}
```

but 

```
object LineBuffer extends App {
  val dWidth = 16
  (new ChiselStage).emitVerilog(
    new LineBuffer(dtype = UInt(dWidth.W), colsIn = 3, rowsOut = 3),
    args
  )
}
```

gives me an exception

```
Elaborating design...
Done elaborating.
Exception in thread "main" firrtl.FirrtlInternalException: Internal Error! trying to emit unsupported operator: UnknownType
Please file an issue at https://github.com/ucb-bar/firrtl/issues
	at firrtl.Utils$.error(Utils.scala:491)
	at firrtl.Utils$.throwInternalError(Utils.scala:164)
	at firrtl.VerilogEmitter.emitCol(VerilogEmitter.scala:212)
	at firrtl.VerilogEmitter.$anonfun$emitCol$2(VerilogEmitter.scala:204)
	at firrtl.VerilogEmitter.$anonfun$emitCol$2$adapted(VerilogEmitter.scala:203)
	at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
	at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
	at scala.collection.immutable.List.foldLeft(List.scala:91)
...
```

### What is the expected behavior?

obviously should be able to run both...

### Steps to Reproduce

is giving you guys the firrtl sufficient to repro?

```
;buildInfoPackage: chisel3, version: 3.4.3, scalaVersion: 2.12.12, sbtVersion: 1.3.10
circuit LineBuffer : 
  module RingBuffer : 
    input clock : Clock
    input reset : Reset
    output io : {flip inData : {valid : UInt<1>, bits : UInt<16>[3]}, outData : UInt<16>[1]}
    
    cmem ram : UInt<16>[3] @[RingBuffer.scala 28:16]
    node _outputPtr_T = eq(io.inData.valid, UInt<1>("h00")) @[RingBuffer.scala 29:32]
    reg outputPtr : UInt<2>, clock with : (reset => (reset, UInt<2>("h00"))) @[Counter.scala 60:40]
    wire outputPtr_wrap : UInt<1>
    outputPtr_wrap <= UInt<1>("h00")
    when _outputPtr_T : @[Counter.scala 118:17]
      node outputPtr_wrap_wrap = eq(outputPtr, UInt<2>("h02")) @[Counter.scala 72:24]
      node _outputPtr_wrap_value_T = add(outputPtr, UInt<1>("h01")) @[Counter.scala 76:24]
      node _outputPtr_wrap_value_T_1 = tail(_outputPtr_wrap_value_T, 1) @[Counter.scala 76:24]
      outputPtr <= _outputPtr_wrap_value_T_1 @[Counter.scala 76:15]
      when outputPtr_wrap_wrap : @[Counter.scala 86:20]
        outputPtr <= UInt<1>("h00") @[Counter.scala 86:28]
        skip @[Counter.scala 86:20]
      outputPtr_wrap <= outputPtr_wrap_wrap @[Counter.scala 118:24]
      skip @[Counter.scala 118:17]
    when io.inData.valid : @[RingBuffer.scala 30:25]
      infer mport MPORT = ram[UInt<1>("h00")], clock @[RingBuffer.scala 32:10]
      MPORT <= io.inData.bits[0] @[RingBuffer.scala 32:14]
      infer mport MPORT_1 = ram[UInt<1>("h01")], clock @[RingBuffer.scala 32:10]
      MPORT_1 <= io.inData.bits[1] @[RingBuffer.scala 32:14]
      infer mport MPORT_2 = ram[UInt<2>("h02")], clock @[RingBuffer.scala 32:10]
      MPORT_2 <= io.inData.bits[2] @[RingBuffer.scala 32:14]
      skip @[RingBuffer.scala 30:25]
    infer mport io_outData_0_MPORT = ram[outputPtr], clock @[RingBuffer.scala 36:23]
    io.outData[0] <= io_outData_0_MPORT @[RingBuffer.scala 36:17]
    
  module RingBuffer_1 : 
    input clock : Clock
    input reset : Reset
    output io : {flip inData : {valid : UInt<1>, bits : UInt<16>[3]}, outData : UInt<16>[1]}
    
    cmem ram : UInt<16>[3] @[RingBuffer.scala 28:16]
    node _outputPtr_T = eq(io.inData.valid, UInt<1>("h00")) @[RingBuffer.scala 29:32]
    reg outputPtr : UInt<2>, clock with : (reset => (reset, UInt<2>("h00"))) @[Counter.scala 60:40]
    wire outputPtr_wrap : UInt<1>
    outputPtr_wrap <= UInt<1>("h00")
    when _outputPtr_T : @[Counter.scala 118:17]
      node outputPtr_wrap_wrap = eq(outputPtr, UInt<2>("h02")) @[Counter.scala 72:24]
      node _outputPtr_wrap_value_T = add(outputPtr, UInt<1>("h01")) @[Counter.scala 76:24]
      node _outputPtr_wrap_value_T_1 = tail(_outputPtr_wrap_value_T, 1) @[Counter.scala 76:24]
      outputPtr <= _outputPtr_wrap_value_T_1 @[Counter.scala 76:15]
      when outputPtr_wrap_wrap : @[Counter.scala 86:20]
        outputPtr <= UInt<1>("h00") @[Counter.scala 86:28]
        skip @[Counter.scala 86:20]
      outputPtr_wrap <= outputPtr_wrap_wrap @[Counter.scala 118:24]
      skip @[Counter.scala 118:17]
    when io.inData.valid : @[RingBuffer.scala 30:25]
      infer mport MPORT = ram[UInt<1>("h00")], clock @[RingBuffer.scala 32:10]
      MPORT <= io.inData.bits[0] @[RingBuffer.scala 32:14]
      infer mport MPORT_1 = ram[UInt<1>("h01")], clock @[RingBuffer.scala 32:10]
      MPORT_1 <= io.inData.bits[1] @[RingBuffer.scala 32:14]
      infer mport MPORT_2 = ram[UInt<2>("h02")], clock @[RingBuffer.scala 32:10]
      MPORT_2 <= io.inData.bits[2] @[RingBuffer.scala 32:14]
      skip @[RingBuffer.scala 30:25]
    infer mport io_outData_0_MPORT = ram[outputPtr], clock @[RingBuffer.scala 36:23]
    io.outData[0] <= io_outData_0_MPORT @[RingBuffer.scala 36:17]
    
  module RingBuffer_2 : 
    input clock : Clock
    input reset : Reset
    output io : {flip inData : {valid : UInt<1>, bits : UInt<16>[3]}, outData : UInt<16>[1]}
    
    cmem ram : UInt<16>[3] @[RingBuffer.scala 28:16]
    node _outputPtr_T = eq(io.inData.valid, UInt<1>("h00")) @[RingBuffer.scala 29:32]
    reg outputPtr : UInt<2>, clock with : (reset => (reset, UInt<2>("h00"))) @[Counter.scala 60:40]
    wire outputPtr_wrap : UInt<1>
    outputPtr_wrap <= UInt<1>("h00")
    when _outputPtr_T : @[Counter.scala 118:17]
      node outputPtr_wrap_wrap = eq(outputPtr, UInt<2>("h02")) @[Counter.scala 72:24]
      node _outputPtr_wrap_value_T = add(outputPtr, UInt<1>("h01")) @[Counter.scala 76:24]
      node _outputPtr_wrap_value_T_1 = tail(_outputPtr_wrap_value_T, 1) @[Counter.scala 76:24]
      outputPtr <= _outputPtr_wrap_value_T_1 @[Counter.scala 76:15]
      when outputPtr_wrap_wrap : @[Counter.scala 86:20]
        outputPtr <= UInt<1>("h00") @[Counter.scala 86:28]
        skip @[Counter.scala 86:20]
      outputPtr_wrap <= outputPtr_wrap_wrap @[Counter.scala 118:24]
      skip @[Counter.scala 118:17]
    when io.inData.valid : @[RingBuffer.scala 30:25]
      infer mport MPORT = ram[UInt<1>("h00")], clock @[RingBuffer.scala 32:10]
      MPORT <= io.inData.bits[0] @[RingBuffer.scala 32:14]
      infer mport MPORT_1 = ram[UInt<1>("h01")], clock @[RingBuffer.scala 32:10]
      MPORT_1 <= io.inData.bits[1] @[RingBuffer.scala 32:14]
      infer mport MPORT_2 = ram[UInt<2>("h02")], clock @[RingBuffer.scala 32:10]
      MPORT_2 <= io.inData.bits[2] @[RingBuffer.scala 32:14]
      skip @[RingBuffer.scala 30:25]
    infer mport io_outData_0_MPORT = ram[outputPtr], clock @[RingBuffer.scala 36:23]
    io.outData[0] <= io_outData_0_MPORT @[RingBuffer.scala 36:17]
    
  module LineBuffer : 
    input clock : Clock
    input reset : UInt<1>
    output io : {flip inData : {valid : UInt<1>, bits : UInt<16>[3]}, outData : {valid : UInt<1>, bits : UInt<16>[3]}}
    
    node _T = eq(io.inData.valid, UInt<1>("h00")) @[LineBuffer.scala 21:45]
    reg _colCntr : UInt<2>, clock with : (reset => (reset, UInt<2>("h00"))) @[Counter.scala 60:40]
    wire colCntrWillWrap : UInt<1>
    colCntrWillWrap <= UInt<1>("h00")
    when _T : @[Counter.scala 118:17]
      node wrap_wrap = eq(_colCntr, UInt<2>("h02")) @[Counter.scala 72:24]
      node _wrap_value_T = add(_colCntr, UInt<1>("h01")) @[Counter.scala 76:24]
      node _wrap_value_T_1 = tail(_wrap_value_T, 1) @[Counter.scala 76:24]
      _colCntr <= _wrap_value_T_1 @[Counter.scala 76:15]
      when wrap_wrap : @[Counter.scala 86:20]
        _colCntr <= UInt<1>("h00") @[Counter.scala 86:28]
        skip @[Counter.scala 86:20]
      colCntrWillWrap <= wrap_wrap @[Counter.scala 118:24]
      skip @[Counter.scala 118:17]
    reg rowCntr : UInt<2>, clock with : (reset => (reset, UInt<2>("h00"))) @[Counter.scala 60:40]
    wire rowCntr_wrap : UInt<1>
    rowCntr_wrap <= UInt<1>("h00")
    when colCntrWillWrap : @[Counter.scala 118:17]
      node rowCntr_wrap_wrap = eq(rowCntr, UInt<2>("h02")) @[Counter.scala 72:24]
      node _rowCntr_wrap_value_T = add(rowCntr, UInt<1>("h01")) @[Counter.scala 76:24]
      node _rowCntr_wrap_value_T_1 = tail(_rowCntr_wrap_value_T, 1) @[Counter.scala 76:24]
      rowCntr <= _rowCntr_wrap_value_T_1 @[Counter.scala 76:15]
      when rowCntr_wrap_wrap : @[Counter.scala 86:20]
        rowCntr <= UInt<1>("h00") @[Counter.scala 86:28]
        skip @[Counter.scala 86:20]
      rowCntr_wrap <= rowCntr_wrap_wrap @[Counter.scala 118:24]
      skip @[Counter.scala 118:17]
    inst RingBuffer of RingBuffer @[LineBuffer.scala 27:49]
    RingBuffer.clock <= clock
    RingBuffer.reset <= reset
    inst RingBuffer_1 of RingBuffer_1 @[LineBuffer.scala 27:49]
    RingBuffer_1.clock <= clock
    RingBuffer_1.reset <= reset
    inst RingBuffer_2 of RingBuffer_2 @[LineBuffer.scala 27:49]
    RingBuffer_2.clock <= clock
    RingBuffer_2.reset <= reset
    wire rings : {flip inData : {valid : UInt<1>, bits : UInt<16>[3]}, outData : UInt<16>[1]}[3] @[LineBuffer.scala 27:22]
    rings[0].outData[0] <= RingBuffer.io.outData[0] @[LineBuffer.scala 27:22]
    RingBuffer.io.inData.bits[0] <= rings[0].inData.bits[0] @[LineBuffer.scala 27:22]
    RingBuffer.io.inData.bits[1] <= rings[0].inData.bits[1] @[LineBuffer.scala 27:22]
    RingBuffer.io.inData.bits[2] <= rings[0].inData.bits[2] @[LineBuffer.scala 27:22]
    RingBuffer.io.inData.valid <= rings[0].inData.valid @[LineBuffer.scala 27:22]
    rings[1].outData[0] <= RingBuffer_1.io.outData[0] @[LineBuffer.scala 27:22]
    RingBuffer_1.io.inData.bits[0] <= rings[1].inData.bits[0] @[LineBuffer.scala 27:22]
    RingBuffer_1.io.inData.bits[1] <= rings[1].inData.bits[1] @[LineBuffer.scala 27:22]
    RingBuffer_1.io.inData.bits[2] <= rings[1].inData.bits[2] @[LineBuffer.scala 27:22]
    RingBuffer_1.io.inData.valid <= rings[1].inData.valid @[LineBuffer.scala 27:22]
    rings[2].outData[0] <= RingBuffer_2.io.outData[0] @[LineBuffer.scala 27:22]
    RingBuffer_2.io.inData.bits[0] <= rings[2].inData.bits[0] @[LineBuffer.scala 27:22]
    RingBuffer_2.io.inData.bits[1] <= rings[2].inData.bits[1] @[LineBuffer.scala 27:22]
    RingBuffer_2.io.inData.bits[2] <= rings[2].inData.bits[2] @[LineBuffer.scala 27:22]
    RingBuffer_2.io.inData.valid <= rings[2].inData.valid @[LineBuffer.scala 27:22]
    io.outData.bits[0] <= UInt<1>("h00") @[LineBuffer.scala 29:24]
    rings[0].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[0].inData.bits[0] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[0].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[0].inData.bits[1] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[0].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[0].inData.bits[2] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    io.outData.bits[1] <= UInt<1>("h00") @[LineBuffer.scala 29:24]
    rings[1].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[1].inData.bits[0] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[1].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[1].inData.bits[1] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[1].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[1].inData.bits[2] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    io.outData.bits[2] <= UInt<1>("h00") @[LineBuffer.scala 29:24]
    rings[2].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[2].inData.bits[0] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[2].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[2].inData.bits[1] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    rings[2].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 31:29]
    rings[2].inData.bits[2] <= UInt<1>("h00") @[LineBuffer.scala 32:31]
    io.outData.valid <= UInt<1>("h00") @[LineBuffer.scala 35:20]
    when io.inData.valid : @[LineBuffer.scala 38:25]
      io.outData.valid <= UInt<1>("h00") @[LineBuffer.scala 39:22]
      rings[rowCntr].inData.valid <= UInt<1>("h01") @[LineBuffer.scala 40:33]
      rings[rowCntr].inData.bits[0] <= io.inData.bits[0] @[LineBuffer.scala 42:37]
      rings[rowCntr].inData.bits[1] <= io.inData.bits[1] @[LineBuffer.scala 42:37]
      rings[rowCntr].inData.bits[2] <= io.inData.bits[2] @[LineBuffer.scala 42:37]
      skip @[LineBuffer.scala 38:25]
    else : @[LineBuffer.scala 44:15]
      io.outData.valid <= UInt<1>("h01") @[LineBuffer.scala 45:22]
      rings[0].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 47:29]
      rings[1].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 47:29]
      rings[2].inData.valid <= UInt<1>("h00") @[LineBuffer.scala 47:29]
      skip @[LineBuffer.scala 44:15]
    node _idx_T = cvt(rowCntr) @[LineBuffer.scala 61:31]
    node _idx_T_1 = sub(_idx_T, asSInt(UInt<4>("h00"))) @[LineBuffer.scala 61:34]
    node _idx_T_2 = tail(_idx_T_1, 1) @[LineBuffer.scala 61:34]
    node _idx_T_3 = asSInt(_idx_T_2) @[LineBuffer.scala 61:34]
    node _idx_T_4 = rem(_idx_T_3, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:67]
    node _idx_T_5 = add(_idx_T_4, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:106]
    node _idx_T_6 = tail(_idx_T_5, 1) @[LineBuffer.scala 61:106]
    node _idx_T_7 = asSInt(_idx_T_6) @[LineBuffer.scala 61:106]
    node _idx_T_8 = rem(_idx_T_7, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:145]
    node idx = asUInt(_idx_T_8) @[LineBuffer.scala 61:190]
    node _T_1 = bits(idx, 1, 0)
    io.outData.bits[_T_1] <= rings[0].outData[0] @[LineBuffer.scala 63:26]
    node _idx_T_9 = cvt(rowCntr) @[LineBuffer.scala 61:31]
    node _idx_T_10 = sub(_idx_T_9, asSInt(UInt<4>("h01"))) @[LineBuffer.scala 61:34]
    node _idx_T_11 = tail(_idx_T_10, 1) @[LineBuffer.scala 61:34]
    node _idx_T_12 = asSInt(_idx_T_11) @[LineBuffer.scala 61:34]
    node _idx_T_13 = rem(_idx_T_12, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:67]
    node _idx_T_14 = add(_idx_T_13, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:106]
    node _idx_T_15 = tail(_idx_T_14, 1) @[LineBuffer.scala 61:106]
    node _idx_T_16 = asSInt(_idx_T_15) @[LineBuffer.scala 61:106]
    node _idx_T_17 = rem(_idx_T_16, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:145]
    node idx_1 = asUInt(_idx_T_17) @[LineBuffer.scala 61:190]
    node _T_2 = bits(idx_1, 1, 0)
    io.outData.bits[_T_2] <= rings[1].outData[0] @[LineBuffer.scala 63:26]
    node _idx_T_18 = cvt(rowCntr) @[LineBuffer.scala 61:31]
    node _idx_T_19 = sub(_idx_T_18, asSInt(UInt<4>("h02"))) @[LineBuffer.scala 61:34]
    node _idx_T_20 = tail(_idx_T_19, 1) @[LineBuffer.scala 61:34]
    node _idx_T_21 = asSInt(_idx_T_20) @[LineBuffer.scala 61:34]
    node _idx_T_22 = rem(_idx_T_21, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:67]
    node _idx_T_23 = add(_idx_T_22, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:106]
    node _idx_T_24 = tail(_idx_T_23, 1) @[LineBuffer.scala 61:106]
    node _idx_T_25 = asSInt(_idx_T_24) @[LineBuffer.scala 61:106]
    node _idx_T_26 = rem(_idx_T_25, asSInt(UInt<4>("h03"))) @[LineBuffer.scala 61:145]
    node idx_2 = asUInt(_idx_T_26) @[LineBuffer.scala 61:190]
    node _T_3 = bits(idx_2, 1, 0)
    io.outData.bits[_T_3] <= rings[2].outData[0] @[LineBuffer.scala 63:26]
```
?

stepping through the debugger inside `VerilogEmitter` i can see that the module writer gets as far as writing 

```
  wire [2:0] _idx_T = {1'b0,$signed(rowCntr)}; // @[LineBuffer.scala 61:31]
  wire [3:0] _GEN_40 = {{1{_idx_T[2]}},_idx_T}; // @[LineBuffer.scala 61:34]
  wire [3:0] _idx_T_3 = $signed(_GEN_40) - 4'sh0; // @[LineBuffer.scala 61:34]
  wire [3:0] _GEN_0 = $signed(_idx_T_3) % 4'sh3; // @[LineBuffer.scala 61:67]
  wire [3:0] _idx_T_4 = _GEN_0[3:0]; // @[LineBuffer.scala 61:67]
  wire 
```

and thus i'm thinking the offending chisel is [here](https://github.com/makslevental/braggnn_fpga/blob/main/src/main/scala/myutil/LineBuffer.scala#L45)

```
  for (r <- 0 until rowsOut) {
    // @formatter:off
    val idx = ((((rowCntr.zext() - r.S((2 * rowCntr.getWidth).W)) % rowsOut.S((2 * rowCntr.getWidth).W)) + rowsOut.S((2 * rowCntr.getWidth).W)) % rowsOut.S((2 * rowCntr.getWidth).W)).asUInt()
    // @formatter:on
    io.outData.bits(idx) := rings(
      r
    ).outData(0)
  }
```

i'm sure this looks weird but that aside i'm not sure why firrtl would work but not verilog.

### Your environment

here's my `build.sbt`

```
// See README.md for license details.

ThisBuild / scalaVersion := "2.12.13"
ThisBuild / version := "0.1.0"
ThisBuild / organization := "com.github.makslevental"

lazy val root = (project in file("."))
  .settings(
    name := "braggnn_chisel",
    libraryDependencies ++= Seq(
      "edu.berkeley.cs" %% "chisel3" % "3.4.3",
      "edu.berkeley.cs" %% "chiseltest" % "0.3.2" % "test",
      "org.scala-lang" % "scala-reflect" % scalaVersion.value,
      "org.scala-lang" % "scala-compiler" % scalaVersion.value,
      "edu.berkeley.cs" %% "chisel-iotesters" % "1.5.3",
      "org.scalanlp" %% "breeze" % "1.1",
      //"edu.berkeley.cs" %% "hardfloat" % "1.3" % "SNAPSHOT",
      "org.scalactic" %% "scalactic" % "3.2.9",
      "org.scalatest" %% "scalatest" % "3.2.9" % "Test", // the capital T is apparently super important???
      "org.rogach" %% "scallop" % "4.0.3",
      "com.lihaoyi" %% "sourcecode" % "0.2.7"
    ),
    scalacOptions ++= Seq(
      "-Xsource:2.11",
      "-language:reflectiveCalls",
      "-deprecation",
      "-feature",
      "-Xcheckinit",
      // Enables autoclonetype2 in 3.4.x (on by default in 3.5)
      "-P:chiselplugin:useBundlePlugin"
    ),
    addCompilerPlugin(("edu.berkeley.cs" % "chisel3-plugin" % "3.4.3").cross(CrossVersion.full)),
    addCompilerPlugin(("org.scalamacros" % "paradise" % "2.1.1").cross(CrossVersion.full))
  )
```

not sure if it matters but jvm is 

```
java 14.0.1 2020-04-14
Java(TM) SE Runtime Environment (build 14.0.1+7)
Java HotSpot(TM) 64-Bit Server VM (build 14.0.1+7, mixed mode, sharing)
```

**Comment by seldridge** [comment]
Thanks for reporting this!

I checked and this is _not_ failing on `1.4.2`, but is failing on `1.4.3` and `1.5-SNAPSHOT` (i.e., `master` branch). So, some regression since `1.4.2`. 

#2129 shows the exact same phenotype and may be provide a smaller test case.

**Comment by seldridge** [comment]
And to clarify, you're not doing anything wrong here, this is a FIRRTL bug. 😬

**Comment by makslevental** [comment]
@seldridge thanks for the quick response. i'll try to rollback to 1.4.2 to get this to work (though i'm betting that'll break something else lol) but looking at https://github.com/chipsalliance/firrtl/issues/2203 am i correct in my reading that i'm violating that constraint on `%` and that that's the source of the issue?

**Comment by seldridge** [comment]
Your Chisel code isn't doing anything wrong here as far as I can tell.

FIRRTL IR doesn't have type information on nodes in it's textual representation (except for on ports). However, this necessarily exists in the in-memory representation in the compiler. In the compiler this works by conservatively assigning anything that it can't trivially figure out as `UnknownType`. There is then a pass, `InferTypes`, which populates the type. 

A bug was introduce after 1.4.2 that is causing some pass to generate an `UnknownType`. This can be fixed by either fixing the pass to not do this or adding an "invalidation" to that pass on `InferTypes`. The compiler will then know that that pass "invalidates" `InferTypes` (translation: the pass may generate `UnknownType`) and the compiler needs to re-run `InferTypes` if any downstream pass states that it needs type information.

The Verilog emitter is then barfing when it sees an `UnknownType`.

tl;dr: The Scala FIRRTL compiler treats type information like an analysis and a pass isn't stating that it invalidates type information.
